plugins/snowflake/dbt/include/snowflake/macros/adapters.sql

```diff
@@ -1,14 +1,28 @@
 {% macro snowflake__create_table_as(temporary, relation, sql) -%}
   {%- set transient = config.get('transient', default=true) -%}
+  {# TODO: Add support for more than one keys #}
+  {%- set cluster_by_keys = config.get('cluster_by') -%}
 
   create or replace {% if temporary -%}
     temporary
   {%- elif transient -%}
     transient
   {%- endif %} table {{ relation }}
   as (
-    {{ sql }}
+    {%- if cluster_by_keys is not none -%}
+      select * from( {{ sql }} ) order by ({{ cluster_by_keys }})
+    {%- else -%}
+      {{ sql }}
+    {%- endif -%}
   );
+
+
+{% endmacro %}
+
+{% macro snowflake__alter_cluster(relation, sql) -%}
+  {# TODO: Add support for more than one keys #}
+  {%- set cluster_by_keys = config.get('cluster_by') -%}
+    alter table {{relation}} cluster by ({{cluster_by_keys}})
```

Comment:
This was news to me when I learned it: altering the table's clustering keys doesn't actually recluster the table. You'd either need to run `alter table ... recluster` (which is now deprecated), or you'd need to enable automatic reclustering for these changes to take effect

Reply:
Damn that's awkward. Indeed I just checked the status of the test table I created and `automatic_clustering` is off.

Somehow, the documentation says there's no way to turn it on. You can suspend it for a table but I do not see a command to enable it.

This kind of leaves us with the following option:
- mimic clustering on tables that you create with order by.
- for incremental tables hope you're using a fairly monotonic cluster by key because most likely Snowflake won't ever to any housekeeping for you later on.
Is that correct?

Reply:
I didn't realize that either! I thought automatic clustering needed to be enabled explicitly.

So, that's kind of good news for us - it means that as long as we specify the clustering keys for the table, Snowflake will automatically recluster it as needed. This should never happen for `table` models since dbt doesn't run DML against those models.

For `incremental` models, each subsequent invocation of dbt should trigger a recluster of the table, which sounds like the right behavior to me!

In my experience, automatic reclustering can be pretty expensive! It might be worth adding a config to suspend automatic reclustering [as shown in the docs](https://docs.snowflake.net/manuals/user-guide/tables-auto-reclustering.html#suspending-automatic-clustering-for-a-table). I'm not exactly sure if there's a use case for setting clustering keys and also suspending automatic reclustering... I think this is an area where I'd need to learn a little bit more, or we'd need to pull in a real expert (like @jtcohen6) :)

Reply:
No no, you were right and maybe I expressed myself badly. Atomatic clustering **does** need to be called explicitly to have Snowflake monitor and perform housekeeping-like reclustering from what I understand and have seen in operation on our db.

I have no idea what happens, if you give clustering keys, but leave automatic reclustering off. Would it just trigger on DMLs? In that case, yeah we don't need to worry. We coulda alter the keys, and bank on the DMLs to trigger reclustering.

We have a bunch of tables on which we have automatic clustering on. Clustered by `created_at` kind of timestamps, and in which we load incrementally. At first run automatic clustering was "expensive" as it was the first run, but on subsequent updates it got a lot cheaper since it only had to do incremental housekeeping. Or at least that's what we see. But my feeling is that if autoclustering is off, having a cluster by key isn't going to do anything which means we'd have to also make sure autoclustering is turned on in the tables that we want it on. 

But yeah I have no idea if I'm right and I'd love to hear from someone who knows Snowflake in and out a bit more.

Reply:
to be honest re reading both the https://docs.snowflake.net/manuals/user-guide/tables-clustering-keys.html and https://docs.snowflake.net/manuals/user-guide/tables-auto-reclustering.html confuses me. Somehow it's not clear whether you'll ever need automatic clustering but at the same time it seems strange that snowflake would keep things clustered without it turned on...

Reply:
@drewbanin I just checked with my boss who has been using clustering a bit in Snowflake and his take on this is that, if you want to benefit from incremental re clustering, you need to specify the clustering keys and turn on reclustering on the table. The order by will take care of loading the data in the right order but if you do any further loading and you have clustering keys such as `user_id`, `created_at` incremental loads wouldn't take care of making the `user_id` be correctly placed so your data would become progressively un clustered and you'd have to do a full-refresh with the order by to re arrange things in the right place.

I just made some changes to reflect this logic. Note that we only need to turn on clustering on incremental tables. For simple tables since they are always fully refreshed the order by is sufficient.

Let me know if you find out conflicting information but, to me, what I did seems to make sense.

Reply:
Sorry for the delay in weighing in! I don't think I'm going to say anything novel, but in the process of saying it, I lend my endorsement to this work, and significant appreciation to @bastienboutonnet.

I may be misremembering—because the last time I dived deeply into this was several months ago when Manual Reclustering was un-deprecated and Automatic Reclustering was still a preview feature—but I believe my memory echoes @drewbanin here:

> it means that as long as we specify the clustering keys for the table, Snowflake will automatically recluster it as needed.

If Automatic Reclustering is enabled for a Snowflake _account_, as soon as a table is defined with cluster keys, Snowflake will start keeping tabs on it.

> This should never happen for table models since dbt doesn't run DML against those models.

We found out that _both_ DML (inserting/updating/deleting rows from an existing clustered table) _and_ DDL (creating a new table with cluster keys) could trigger automatic clustering, if Snowflake identified that the newly created table was not optimally clustered, based on the keys defined for it `post hoc`. This ended up being the primary expense of our experimentation :) and it's avoided, as this PR does, by creating the tables with an equivalent `order by`. That's this line from [the top of the docs](https://docs.snowflake.net/manuals/user-guide/tables-auto-reclustering.html):

> Note that, after a clustered table is defined, reclustering does not necessarily start immediately. Snowflake only reclusters a clustered table if it will benefit from the operation.

The correctly ordered table has nothing to benefit from the operation; automatic clustering is still _enabled_, but finds itself with nothing to do.

> For incremental models, each subsequent invocation of dbt should trigger a recluster of the table, which sounds like the right behavior to me!

Right on. As long as the original table is properly ordered, as are any subsequent temp tables for merging, the subsequent automatic clustering can be a fairly performant, fairly cheap interleaving.

So my only practical thought here is that, per my understanding of late 2018, the `automatic_clustering` config and `alter table {{relation}} resume recluster` DML is not actually needed. When automatic clustering is turned on for the account, any table with cluster keys defined is presumed to have auto-clustering _resumed_ until it is explicitly _suspended_.

That said, it's a nice idea for dbt to be verbose-and-beyond in its operations, especially ones this subtle, and so I've similarly included the post-hook in past implementations.

Reply:
Thanks for your reply @jtcohen6 <3 . I think your comments make sense and indeed confirm most of the things we wanted to check. Regarding turning clustering on at the table level (after of course having it enabled for the account for it to work) we found that if we do not run it, the `automatic_clustering` column for a table for which we have given clustering keys is marked as `OFF` which suggests that automatic clustering may not be performed as you suggest. I'm curious how you had tested it previously. 

As you say we don't risk a lot by being extra explicit and I actually like to be extra explicit but I'm curious to really make sure I understand before just hiding behind a "it won't hurt" logic. Always better not to implement stuff that's un-necessary.

Reply:
@bastienboutonnet That could be! That morsel of knowledge was one I had picked up when this feature was in private release many months ago, and it struck me as odd at the time that `automatic_clustering` _would_ be enabled on tables, by default, as soon as a cluster key was defined. So I'm not surprised and actually pleased to hear that the switch has flipped.

Snowflake's docs on the matter are still ambiguous IMO, so I think we err well on the side of being extra un-ambiguous.

Reply:
@bastienboutonnet That could be! That morsel of knowledge was one I had picked up when this feature was in private release many months ago, and it struck me as odd at the time that automatic_clustering would be enabled on tables, by default, as soon as a cluster key was defined. So I'm not surprised and actually pleased to hear that the switch has flipped.

Snowflake's docs on the matter are still ambiguous IMO, so I think we err well on the side of being extra un-ambiguous.

Reply:
Awesome! then I guess it all makes sense. Yes the documentation around that feature is hella ambiguous. I should raise it as an issue to our technical rep or on the snowflake community as I think it probably leaves more than us confused.